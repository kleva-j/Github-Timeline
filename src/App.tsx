import * as React from 'react';
import { Redirect, Route, Switch } from 'react-router-dom';

/* Import Components */
import { ProtectedRoute } from './components/protected.route';

/* Import Views */
import { Timeline } from './views/Timeline';
import { Login } from './views/Login';
import { Jobs } from './views/Jobs';

const App: React.FC = () => {
  return (
    <>
      {/* Page header goes here!!! */}
      <Switch>
        <Redirect exact from="/" to="/timeline" />
        <ProtectedRoute exact path="/timeline" component={Timeline} />
        <Route exact path="/jobs" component={Jobs} />
        <Route exact path="/login" component={Login} />
        <Route path="*" component={() => <h1>404 NOT FOUND</h1>} />
      </Switch>
    </>
  );
};

export default App;
