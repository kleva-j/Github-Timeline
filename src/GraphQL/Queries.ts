import { gql } from '@apollo/client';

export const LOAD_JOBS = gql`
  query GetJobs {
    jobs {
      id
      title
      type
      company
      description
      url
      location
      created_at
    }
  }
`;

export const LOAD_SINGLE_JOB = gql`
  query GetJob($id: ID!) {
    job(id: $id) {
      id
      title
      type
      company
      description
      url
      location
      created_at
    }
  }
`;
