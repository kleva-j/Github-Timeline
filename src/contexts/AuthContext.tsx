import React, { createContext, useContext, useReducer } from 'react';

import { AuthReducer, initialState, IState } from 'reducers/auth';

export const AuthStateContext = createContext<IState>(initialState);
export const AuthDispatchContext = createContext<React.Dispatch<any> | null>(null);

export const useAuthState = () => {
  const context = useContext(AuthStateContext);
  // console.log(context, 'Check here')
  if (context === undefined) {
    throw new Error('useAuthState must be used within a AuthProvider');
  }
  return context;
};

export const useAuthDispatch = () => {
  const context = useContext(AuthDispatchContext);
  if (context === undefined) {
    throw new Error('useAuthDispatch must be used within a AuthProvider');
  }
  return context;
};

export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [user, dispatch] = useReducer(AuthReducer, initialState);

  return (
    <AuthStateContext.Provider value={user}>
      <AuthDispatchContext.Provider value={dispatch}>{children}</AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  );
};
