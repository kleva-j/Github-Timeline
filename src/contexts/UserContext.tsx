import React, { createContext, useReducer, useContext } from 'react';

import { initialState, UserReducer } from 'reducers/user';

export interface UserInterface {
  currentUser: User | undefined;
  loading: boolean;
  isAuthenticated: boolean;
  error: Error | null;
  theme?: any;
}

type User = {
  username: string | undefined;
};

export const UserStateContext = createContext<UserInterface>(initialState);
export const UserDispatchContext = createContext<React.Dispatch<any> | null>(null);

export const useUserState = () => {
  const context = useContext(UserStateContext);
  if (context === undefined) {
    throw new Error('useUserState must be used within a UserProvider');
  }
  return context;
};

export const useUserDispatch = () => {
  const context = useContext(UserDispatchContext);
  if (context === undefined) {
    throw new Error('useUserDispatch must be used within a UserProvider');
  }
  return context;
};

export const UserProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [user, dispatch] = useReducer(UserReducer, initialState);

  return (
    <UserStateContext.Provider value={user}>
      <UserDispatchContext.Provider value={dispatch}>{children}</UserDispatchContext.Provider>
    </UserStateContext.Provider>
  );
};
