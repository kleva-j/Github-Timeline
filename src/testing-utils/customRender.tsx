import { render } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';

export const waitForData = () => new Promise((res) => setTimeout(res, 0));

export const customRender = (node: JSX.Element, mocks?: MockedResponse[]) => {
  return {
    ...render(
      <MockedProvider mocks={mocks} addTypename={false}>
        {node}
      </MockedProvider>,
    ),
  };
};
