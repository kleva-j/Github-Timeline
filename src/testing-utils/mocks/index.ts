import { LOAD_JOBS, LOAD_SINGLE_JOB } from 'GraphQL/Queries';

export const mockSingleJob = {
  request: {
    query: LOAD_SINGLE_JOB,
    variables: { id: 'XXXX' },
  },
  result: {
    data: {
      job: {
        id: 'XXXX',
        title: 'Buck',
        type: 'bulldog',
        company: 'company',
        description: 'description',
        url: 'some_random_url',
        location: 'location',
        created_at: 'Fri Mar 12 09:57:21 UTC 2021',
        company_url: 'some_company_url',
        company_logo: 'some_company_logo',
        how_to_apply: 'how_to_apply',
      },
    },
  },
};

export const errorMockSingleJob = {
  request: mockSingleJob.request,
  error: new Error('An error occurred'),
};

export const mockAllJobs = {
  request: {
    query: LOAD_JOBS,
    variables: {},
  },
  result: {
    data: {
      jobs: [
        {
          id: 'XXXX',
          title: 'Buck',
          type: 'bulldog',
          company: 'company',
          description: 'description',
          url: 'some_random_url',
          location: 'location',
          created_at: 'Fri Mar 12 09:57:21 UTC 2021',
          company_url: 'some_company_url',
          company_logo: 'some_company_logo',
          how_to_apply: 'how_to_apply',
        },
      ],
    },
  },
};

export const errorMockAllJobs = {
  request: mockAllJobs.request,
  error: new Error('An error occurred'),
};
