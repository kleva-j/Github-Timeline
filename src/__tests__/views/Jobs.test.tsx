import { MemoryRouter } from 'react-router-dom';
import { cleanup } from '@testing-library/react';

import App from 'App';
import { LOAD_JOBS } from 'GraphQL/Queries';
import { AuthProvider } from 'contexts/AuthContext';
import { customRender, waitForData } from 'testing-utils/customRender';

const mock = {
  request: {
    query: LOAD_JOBS,
    variables: {},
  },
  result: {
    data: {
      jobs: [
        {
          id: 'XXXX',
          title: 'Buck',
          type: 'bulldog',
          company: 'company',
          description: 'description',
          url: 'some_random_url',
          location: 'location',
          created_at: 'created_at',
        },
      ],
    },
  },
};

describe('Jobs Page', () => {
  afterEach(cleanup);

  test('should render the Jobs page', async () => {
    const { container } = customRender(
      <AuthProvider>
        <MemoryRouter initialEntries={['/jobs']}>
          <App />
        </MemoryRouter>
      </AuthProvider>,
      [mock],
    );
    await waitForData();
    expect(container.firstChild).toMatchSnapshot();
  });
});
