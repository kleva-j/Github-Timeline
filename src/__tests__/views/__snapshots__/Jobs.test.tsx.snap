// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jobs Page should render the Jobs page 1`] = `
<section>
  This is the Jobs page.
</section>
`;
